socorro/external/boto/connection_context.py

```diff
@@ -14,20 +14,72 @@
 from configman import Namespace, RequiredConfig, class_converter
 
 from socorrolib.lib.converters import change_default
+from socorrolib.lib.ooid import dateFromOoid
 
 
 class KeyNotFound(Exception):
     pass
 
 
 class JSONISOEncoder(json.JSONEncoder):
-
     def default(self, obj):
         if isinstance(obj, datetime.date):
             return obj.isoformat()
         raise NotImplementedError("Don't know about {0!r}".format(obj))
 
 
+class KeyBuilderBase(object):
+    """Base key builder for s3 pseudo-filenames"""
+    def build_keys(self, prefix, name_of_thing, id, version='v1'):
+        """
+        Use S3 pseudo-directories to make it easier to list/expire later
+        {prefix}/{version}/{name_of_thing}/{id}
+        """
+        return [
+            '%s/%s/%s/%s' % (prefix, version, name_of_thing, id)
+        ]
+
+
+class DatePrefixKeyBuilder(KeyBuilderBase):
+    """s3 pseudo-filename key builder with date prefixes
+
+    This adds the date to the prefix for raw_crash objects making it easier to
+    list all the raw_crash objects for a specific date.
+
+    """
+    def build_keys(self, prefix, name_of_thing, id, version='v1'):
+        """Use S3 pseudo-directories to make it easier to list/expire.
+
+        For "raw_crash" things, this pulls out a date and adds that as an
+        additional part of the directory structure::
+
+            {prefix}/{version}/{name_of_thing}/{date}/{id}
+
+        Then it also returns the old-style key so that we can fetch objects
+        that were saved with the old key::
+
+            {prefix}/{version}/{name_of_thing}/{id}
+
+        """
+        keys = []
+
+        if name_of_thing == 'raw_crash':
+            datestamp = dateFromOoid(id)
+
+            if datestamp is not None:
+                date = datestamp.strftime('%Y%m%d')
```

Comment:
I switch it to a 4-digit year because that's what I think that's the format of the least surprise.

Does that sound right? Is there a precedent for 2-digit years in Socorro other than the ooid which has space restrictions?